### DISCUSSION TITLE
Reduce Executable Size -- Multiple Copies of OpenBLAS?

### DISCUSSION THREAD

**Discussion by hanzhi713** [pyinstaller/pyinstaller#6389] [Help] [answered]
Hi, I'm trying to build executables for my application on github actions. My application depends on numpy and scipy. The executable sizes on macos and linux are very large, and the main cause it seems is that multiple copies OpenBLAS (and some other shared libraries) are in the built executable. The screenshot showed the duplicated BLAS libraries from the macos build. 

![image](https://user-images.githubusercontent.com/21034047/143726350-62eef8e8-484d-4c6d-b4c4-e004a26727e1.png)

The github action workflow can be found here: https://github.com/hanzhi713/image-collage-maker/blob/master/.github/workflows/build_executable.yaml
I've also uploaded executables and the build folders (those end with `-build`) for reference in a temporary release: https://github.com/hanzhi713/image-collage-maker/releases/tag/vtest

Is there a way to eliminate redundant copies of the same libraries? Or are there other ways to reduce the executable size?

**Comment by bwoodsend** [comment, marked as answer]
Looking at the `build/photomosaic-maker-3.0-macos-x64-archive/Analysis-00.toc`, I see that there are duplicate copies of libblas in there because Conda ships duplicate copies of libblas. This is one of the many reasons I keep telling people to **stop using Conda**!

**Reply by rokm** [reply]
I suppose originally those might be symbolic links to a single "real" shared library. But because other stuff ends up referencing them by different names, we need to collect them all (or we end up collecting them all due to `numpy` hook). And because we don't support symbolic links, we end up with multiple copies of the same file, under different names.

The only way around this until we implement proper (or at least provisional) support for symbolic links, is to manually remove the redundant copies and replace them with symbolic links (which works only for `onedir` builds, where you can do manual post-processing).

**Reply by rokm** [reply]
Which is indeed the case:
```
$ ls -l /Users/rok/opt/anaconda3/envs/test-pyi/lib/*blas*
lrwxr-xr-x  1 rok  staff        26 28 nov 12:47 /Users/rok/opt/anaconda3/envs/test-pyi/lib/libblas.dylib -> libopenblasp-r0.3.13.dylib
lrwxr-xr-x  1 rok  staff        26 28 nov 12:47 /Users/rok/opt/anaconda3/envs/test-pyi/lib/libcblas.dylib -> libopenblasp-r0.3.13.dylib
lrwxr-xr-x  1 rok  staff        26 28 nov 12:47 /Users/rok/opt/anaconda3/envs/test-pyi/lib/libopenblas.0.dylib -> libopenblasp-r0.3.13.dylib
lrwxr-xr-x  1 rok  staff        26 28 nov 12:47 /Users/rok/opt/anaconda3/envs/test-pyi/lib/libopenblas.dylib -> libopenblasp-r0.3.13.dylib
-rwxrwxr-x  1 rok  staff  65000816 28 nov 12:47 /Users/rok/opt/anaconda3/envs/test-pyi/lib/libopenblasp-r0.3.13.dylib
```

Another problem I see from the original list is that `libopenblas.dylib` ended up different than others; probably it was properly collected as `BINARY` and underwent library path rewriting and other magic, whereas other instances were collected as `DATA` and did not...

**Reply by rokm** [reply]
Now that I think about it, having multiple copies of the library might also result in crashes if other stuff indeed ends up referencing them by different names, similarly to #5710.

**Reply by hanzhi713** [reply]
I switched to venv to create build environments and it worked great. The executable sizes are significantly reduced (see https://github.com/hanzhi713/image-collage-maker/releases/tag/v3.1). Thanks for the help.
